networkx/algorithms/swap.py

```diff
@@ -6,7 +6,130 @@
+    a -> c -> b -> d. This pattern of swapping allows all possible states with the
+    same in- and out-degree distribution in a directed graph to be reached.
+
+    If the swap would create parallel edges (e.g. if a -> c already existed in the
+    previous example), another attempt is made to find a suitable trio of edges.
+
+    Parameters
+    ----------
+    G : DiGraph
+       A directed graph
+
+    nswap : integer (optional, default=1)
+       Number of three-edge swaps to perform
+
+    max_tries : integer (optional)
+       Maximum number of attempts to swap edges
+
+    seed : integer, random_state, or None (default)
+        Indicator of random number generation state.
+        See :ref:`Randomness<randomness>`.
+
+    Returns
+    -------
+    G : DiGraph
+       The graph after triple-edge swaps.
+
+    Raises
+    ------
+    NetworkXError
+        If `G` is not directed, or
+        If nswap > max_tries, or
+        If there are fewer than 4 nodes in `G`
+    NetworkXAlgorithmError
+        If the number of swap attempts exceeds `max_tries` before `nswap` swaps are made
+
+    Notes
+    -----
+    Does not enforce any connectivity constraints.
+
+    The graph G is modified in place.
+
+    References
+    ----------
+    .. [1] Erdős, Péter L., et al. “A Simple Havel-Hakimi Type Algorithm to Realize
+           Graphical Degree Sequences of Directed Graphs.” ArXiv:0905.4913 [Math],
+           Jan. 2010. https://doi.org/10.48550/arXiv.0905.4913.
+    .. [2] “Combinatorics - Reaching All Possible Simple Directed Graphs with a given
+           Degree Sequence with 2-Edge Swaps.” Mathematics Stack Exchange,
+           https://math.stackexchange.com/questions/22272/. Accessed 30 May 2022.
+    """
+    if not G.is_directed():
+        raise nx.NetworkXError(
+            "directed_edge_swap() is only defined for directed graphs. Use double_edge_swap "
+            "instead"
+        )
+    if nswap > max_tries:
+        raise nx.NetworkXError("Number of swaps > number of tries allowed.")
+    if len(G) < 4:
+        raise nx.NetworkXError("Graph has less than four nodes.")
+
+    # Instead of choosing uniformly at random from a generated edge list,
+    # this algorithm chooses nonuniformly from the set of nodes with
+    # probability weighted by degree.
+    n = 0
+    swapcount = 0
+    keys, degrees = zip(*G.degree())  # keys, degree
+    cdf = nx.utils.cumulative_distribution(degrees)  # cdf of degree
+    discrete_sequence = nx.utils.discrete_sequence
+
+    while swapcount < nswap:
+        # choose source node index from discrete distribution
+        start_index = discrete_sequence(1, cdistribution=cdf, seed=seed)[0]
+        start = keys[start_index]
+
+        # If the given node doesn't have any out edges, then there isn't anything to swap
+        if G.out_degree(start) == 0:
+            n += 1
+            continue
+        _, second = seed.choice(list(G.out_edges(start)))
```

Comment:
This idiom uses `out_edges to find successors. It would be better to use the `list(G.succ[start])` idiom.  Or maybe better, since this lookup could be done quite often, a preprocessing step could build a dict of successor lists. That would avoid constructing the same list many times:  `nbrs = {n: list(G.succ[n]) for n in G}` and then 1st second = seed.choice(nbrs[start])`.  Similarly for the other choice lines.

Reply:
I updated to use the `G.succ` idiom, but avoided prebuilding the lists for this and the following comment. 

My first thought was that in situations with many nodes and few swaps, prebuilding the list would be slow and a cache would be preferred. I started to implement a cache, but realized that most of the time when we get a cache hit we'll immediately have to rebuild the list for that cache entry anyway since we're modifying that node's successors. 

I think for sparse graphs or graphs with a lot of isolated cycles it might still be worth implementing a cache, but for graphs where most iterations result in a successful swap it's a large amount of memory overhead exchanged for little to no runtime speedup

Reply:
You are right that caching or (prebuilding) the lists will not be effective here because the graph is being morphed so the overhead of updating the lists would likely be more than simply rebuilding it when needed. :}